### ISSUE TITLE
[Info] Memory reclaim policy in Android OS

### ISSUE THREAD

**Issue by periannath** [Samsung/ONE#5461]
This issue is to share a recent finding of Android memory management.

### Tl;dr Android application delays reclaim of freed memory.

While testing on android application, I found that freed memory is not claimed by OS immediately. However, in android console, freed memory  is immediately claimed by OS.

I found a document for Android native memory allocator while searching. https://android.googlesource.com/platform/bionic/+/master/docs/native_allocator.md

Here is the feature related to our problem.

#### M_DECAY_TIME

> When set to zero, mallopt(M_DECAY_TIME, 0), it is expected that an allocator will attempt to purge and release any unused memory back to the kernel on free calls. This is important in Android to avoid consuming extra RSS.
>
> When set to non-zero, mallopt(M_DECAY_TIME, 1), an allocator can delay the purge and release action. The amount of delay is up to the allocator implementation, but it should be a reasonable amount of time. The jemalloc allocator was implemented to have a one second delay.
>
> The drawback to this option is that most allocators do not have a separate thread to handle the purge, so the decay is only handled when an allocation operation occurs. For server processes, this can mean that RSS is slightly higher when the server is waiting for the next connection and no other allocation calls are made. The M_PURGE option is used to force a purge in this case.
>
> For all applications on Android, the call mallopt(M_DECAY_TIME, 1) is made by default. The idea is that it allows application frees to run a bit faster, while only increasing RSS a bit.

- Peak memory measured by android app and `nnpackage_run` can be different since they use different reclaim policy. 
- When I ran `nnpackage_run` with `mallopt(M_DECAY_TIME, 1)`, peak memory usage is increased since freed memory is not claimed immediately.

### How to reclaim freed memory immediately for small peak RSS
1. Call `mallopt(M_DECAY_TIME, 0);` once in ONE runtime
   - Immediate reclaim might cause some slowdown
1. Call `mallopt(M_PURGE, 0);` everytime after deallocation
   - `mallopt(M_PURGE, 0);` forces OS to relcaim freed memory
   - Example : https://github.com/Samsung/ONE/pull/5450
